Honors Project, Refactored Saved Regressions and Database Migration, Regression Lines no Longer Bounded by Data

JKinsman opened the pull request:
This patch essentially contains a regression overhaul.

1.  Timeline and Scatter Plot now feature symbolic regressions! :smile: 
2.  Regression lines are no longer bounded by data (zooming out will extend the regression function on the vis to the ends of the x axis).  Addresses issue #1844 
3.  Saved visualizations of all types are refactored to a new standardized format.  This is ideal for implementing item two, in addition to eliminating redundant code during the process of implementing item one.
4.  The. Most. Evil. Database. Migration. Ever. :stuck_out_tongue: 

@jaypoulz , I am so, so sorry... :sob: 

@fgmart

jaypoulz commented:
![wat-regr](https://cloud.githubusercontent.com/assets/1039164/7038916/bb7d43ca-dd86-11e4-84cf-04cece6cdc93.png)

I managed to break this making a linear regression on timeline. For some reason, the line generated is way below the data.

jaypoulz commented on app/assets/javascripts/visualizations/highvis/binaryTree.coffee:
```diff
@@ -0,0 +1,393 @@
+
+      # Returns true if a and b are equivalent objects
+      # (not necessarily references to the same object in memory).
+      @is_equal: (a, b) ->
```
camelCase bro -> '_' delimiters are for ruby code.

jaypoulz commented on app/assets/javascripts/visualizations/highvis/binaryTree.coffee:
```diff
@@ -0,0 +1,393 @@
+        leftEq and rightEq
+
+      # Checks if the tree is terminal (i.e., no children)
+      is_terminal: ->
```
:camel: case

jaypoulz commented on app/assets/javascripts/visualizations/highvis/binaryTree.coffee:
```diff
@@ -0,0 +1,393 @@
+          rest = combiner(rest, @left.__query(combiner))
+        if @right? and @right isnt null
+          rest = combiner(rest, @right.__query(combiner))
+        1 + rest
```
please a return -> it can make things easier to read

jaypoulz commented on app/assets/javascripts/visualizations/highvis/binaryTree.coffee:
```diff
@@ -0,0 +1,393 @@
+      ###
+      insertData: (data, pos = null) ->
+        if pos is 'left'
+          if @data isnt null
```
```
if @data is null
  console.log "Error inserting #{data} into left child of tree."
  return
```
Will prevent a third level of indentation with an implied else

jaypoulz commented on app/assets/javascripts/visualizations/highvis/binaryTree.coffee:
```diff
@@ -0,0 +1,393 @@
+          else
+            console.log "Error inserting #{data} into left child of tree."
+        else if pos is 'right'
+          if @data isnt null
```
use the implied else trick

jaypoulz commented on app/assets/javascripts/visualizations/highvis/binaryTree.coffee:
```diff
@@ -0,0 +1,393 @@
+      ###
+      deleteData: (pos = null) ->
+        if pos is 'right'
+          if @right.is_terminal()
```
implied else and returns

jaypoulz commented on app/assets/javascripts/visualizations/highvis/binaryTree.coffee:
```diff
@@ -0,0 +1,393 @@
+      ###
+      generate: (maxDepth = 10, curDepth = 1) ->
+        if curDepth is maxDepth
+          @insertData(window.BinaryTree.terminals[Math.floor(Math.random() * window.BinaryTree.terminals.length)])
```
you don't need 'window' -> the default scope is window, so you already have access to it

jaypoulz commented on app/assets/javascripts/visualizations/highvis/binaryTree.coffee:
```diff
@@ -0,0 +1,393 @@
+        replacementPoint = @index(index)
+        if index is null or replacementPoint is -1
+          console.log "Error inserting #{tree} at location specified.  Index does not exist in the tree."
+          null
```
return null

jaypoulz commented on app/assets/javascripts/visualizations/highvis/binaryTree.coffee:
```diff
@@ -0,0 +1,393 @@
+        if index is null or replacementPoint is -1
+          console.log "Error inserting #{tree} at location specified.  Index does not exist in the tree."
+          null
+        else
```
this else is implied

jaypoulz commented on app/assets/javascripts/visualizations/highvis/binaryTree.coffee:
```diff
@@ -0,0 +1,393 @@
+        [tree1a, tree1b] = [window.BinaryTree.clone(tree1), window.BinaryTree.clone(tree1)]
+        [tree2a, tree2b] = [window.BinaryTree.clone(tree2), window.BinaryTree.clone(tree2)]
+        
+        [crossoverPointOne, crossoverPointTwo] = \
```
don't use \ to extend the line. just indent the line below once

jaypoulz commented on app/assets/javascripts/visualizations/highvis/binaryTree.coffee:
```diff
@@ -0,0 +1,393 @@
+
+      # Given a tree, construct a string representation of the mathematical
+      # function the tree describes
+      @stringify: (tree) ->
```
Just a note here.

@stringify expands to this.stringify which is really window.stringify which is why this works.

jaypoulz commented on app/assets/javascripts/visualizations/highvis/binaryTree.coffee:
```diff
@@ -0,0 +1,393 @@
+        switch tree.data
+          when window.add
+            "#{parenthesize(@stringify(tree.left))} + #{parenthesize(@stringify(tree.right))}"
+
```
either each case gets spaces or none of them do

jaypoulz commented on app/assets/javascripts/visualizations/highvis/binaryTree.coffee:
```diff
@@ -0,0 +1,393 @@
+          when window.add
+            "#{parenthesize(@stringify(tree.left))} + #{parenthesize(@stringify(tree.right))}"
+
+          when window.subtract
```
you should find replace all of this window.blahs -> they aren't necessary except in the declarator

jaypoulz commented:
I'm not a fan of adding all of these functions directly to window - we are begging for namespace collision. Try starting with the line:
```window.symregr?= {}```

And then replace all of the window.blah stuff with symregr.blah

jaypoulz commented on app/assets/javascripts/visualizations/highvis/individual.coffee:
```diff
@@ -0,0 +1,372 @@
+        
+        # b = cov(y,t) / var(y)
+        b = if yVar is 0 then 1 else ytCov / yVar
+        # a = tAvg - b * yAvg
```
This comment is not helpful

jaypoulz commented on app/assets/javascripts/visualizations/highvis/individual.coffee:
```diff
@@ -0,0 +1,372 @@
+          for particle in particles
+            evaluationPoint = new Individual(tree, populant.maxDepth)
+            for dimension, index in particle.dimensions
+              dimension.velocity = Math.min(Math.max(dimension.velocity + (Math.random() * c1 * \
```
once again do extend lines with \
just indent the next lines appropriately

jaypoulz commented on app/assets/javascripts/visualizations/highvis/regression.coffee:
```diff
@@ -394,4 +421,17 @@ $ ->
+      # Get R2 value
+      [xs, ys] = [(point.x for point in points), (point.y for point in points)]
+      yAvg = ys.reduce((pv, cv, index, array) -> pv + cv) / ys.length
+      ssRes = (Math.pow(y - solution.evaluate(xs[i]), 2) for y, i in ys)\
```
don't extend the line, just newline after a comma and indent

jaypoulz commented on app/assets/javascripts/visualizations/highvis/regression.coffee:
```diff
@@ -394,4 +421,17 @@ $ ->
+
+    calculateR2 = (solution, points) ->
+      # Get R2 value
+      [xs, ys] = [(point.x for point in points), (point.y for point in points)]
```
This is seriously a Grinstein approved line

stowellm replied:
By that Jeremy means the line below this :goat: `for point in points`

jaypoulz commented on app/assets/javascripts/visualizations/highvis/scatter.coffee:
```diff
@@ -314,22 +314,39 @@ $ ->
-          if regression.fieldIndices[0] == @configs.xAxis \
-          && "#{regression.fieldIndices[2]}" == "#{data.groupSelection}" \
-          && globals.configs.fieldSelection.indexOf(regression.fieldIndices[1]) != -1
+          if regression.xAxis is @configs.xAxis \
```
keep the and on the same line, then line break - you won't need the \

jaypoulz commented on app/assets/javascripts/visualizations/highvis/symbolicRegression.coffee:
```diff
@@ -0,0 +1,369 @@
+    ###
+
+    # Fitness function to be used (scaled mean-squared error)
+    window.FITNESS = 'scaledFitness'
```
we need to namespace this stuff -> see my comment about namespacing

jaypoulz commented on app/assets/javascripts/visualizations/highvis/symbolicRegression.coffee:
```diff
@@ -0,0 +1,369 @@
+        Math.max(pv, cv)
+
+      population = (new Individual(null, INDIVIDUALDEPTH) for i in [0...POPULATIONSIZE])
+      #population = population.map((y) -> Individual.particleSwarmOptimization(y, points))
```
comment not useful

jaypoulz commented on db/migrate/20150313212337_refactor_saved_regressions.rb:
```diff
@@ -0,0 +1,114 @@
+            name = regression['series']['name']['group']
+            r2 = regression['series']['name']['regression']['tooltip'].split('</strong> ')[3].gsub('e', 'E')
+            type = regression['type']
+            wtf = regression['series']['name']['regression']['tooltip'].split('<br>')[1].delete('^0-9 \.\-eE').split('  ').reverse
```
lololol @ variable but 'whata terrible failure' should probably not be in our code

JKinsman replied:
Loool, I totally forgot to change that!  Oops! :joy:

jaypoulz commented on db/migrate/20150313212337_refactor_saved_regressions.rb:
```diff
@@ -0,0 +1,114 @@
+        globals['Scatter']['savedRegressions'] = scatter_regressions
+      end
+
+      if !timeline_params.nil? and timeline_params.key? 'savedRegressions'
```
you copy pasted scatter instead of making a function?

I realize this is only a migration but like, seriously we don't have to the write the worst code ever :P

jaypoulz commented:
OK so heres the final deal: I want you to fix the styling in the files you added. Don't bother making the changes in files you edited, because it'll be a nightmare to merge. I'll make those myself.

JKinsman commented:
@jaypoulz I addressed your concerns.

Change summary:

1. Global variables in window moved to window.symregr (avoid namespace collision)
2. General styling (I think I got all of them)
     a. No more line extensions, replaced with proper indentation
     b. Consistent spacing
     c. Removed useless comments
     d. Removed useless window prefixes
3.  Cleaned up the migration code (it's actually reasonable now)
4.  Fixed a bug in parameter parsing -- turns out the regex was capturing exponents and adding them to the end of parameters on quadratic / cubic coefficients.  It also captured the 'e' in exponential regressions.  Ensured all parameters are now parsed out properly.
5.  Fixed regressions on Timeline bug (this involved removing one line of code from Scatter).  
     a.  The line is in the try block under the call to get regression.  It's a map over Ps that was meant to go elsewhere in a previous iteration of the code.  If this will cause you merge conflicts, I can add it back and have you delete it.

I did not make any of your suggested changes in regression/scatter.coffee, per your request.

Thanks Jeremy! :smile:

jaypoulz commented:
@kcarcia could we have someone do a secondary review? I played with this and am mostly satisfied.

kcarcia commented:
@jaypoulz Sure, I can look this over tomorrow morning!

kcarcia commented on app/assets/javascripts/visualizations/highvis/binaryTree.coffee:
```diff
@@ -0,0 +1,374 @@
+      insertData: (data, pos = null) ->
+        if pos is 'left'
+          if @data is null
+            console.log "Error inserting #{data} into left child of tree."
```
Remove or comment out console.log please

kcarcia commented on app/assets/javascripts/visualizations/highvis/binaryTree.coffee:
```diff
@@ -0,0 +1,374 @@
+          @left.data = data
+        else if pos is 'right'
+          if @data is null
+            console.log "Error inserting #{data} into right child of tree."
```
Remove or comment out console.log please

kcarcia commented on app/assets/javascripts/visualizations/highvis/binaryTree.coffee:
```diff
@@ -0,0 +1,374 @@
+          if @right.isTerminal()
+            @right = null
+            return
+          console.log "Error deleting #{@right.data}, results in invalid binary tree."
```
Remove or comment out console.log please

kcarcia commented on app/assets/javascripts/visualizations/highvis/binaryTree.coffee:
```diff
@@ -0,0 +1,374 @@
+          if @left.isTerminal()
+            @left = null
+            return
+          console.log "Error deleting #{@left.data}, results in invalid binary tree."
```
Remove or comment out console.log please

kcarcia commented on app/assets/javascripts/visualizations/highvis/binaryTree.coffee:
```diff
@@ -0,0 +1,374 @@
+            @data = null
+            return
+          console.log "Error deleting #{@data}, results in invalid binary tree."
+          null
```
Remove or comment out console.log please

kcarcia commented on app/assets/javascripts/visualizations/highvis/binaryTree.coffee:
```diff
@@ -0,0 +1,374 @@
+          if @isTerminal()
+            @data = null
+            return
+          console.log "Error deleting #{@data}, results in invalid binary tree."
```
Remove or comment out console.log please

kcarcia commented on app/assets/javascripts/visualizations/highvis/binaryTree.coffee:
```diff
@@ -0,0 +1,374 @@
+      insertTree: (tree, index = 0) ->
+        replacementPoint = @index(index)
+        if index is null or replacementPoint is -1
+          console.log "Error inserting #{tree} at location specified.  Index does not exist in the tree."
```
Remove or comment out console.log please

kcarcia commented on app/assets/javascripts/visualizations/highvis/scatter.coffee:
```diff
@@ -623,8 +640,8 @@ $ ->
-        ($ '#controldiv').append controls
-        ($ "#regressionControl button").button()
+        $('#controldiv').append controls
+        $("#regressionControl button").button()
```
single quotes pease

kcarcia commented on app/assets/javascripts/visualizations/highvis/scatter.coffee:
```diff
@@ -632,178 +649,207 @@ $ ->
           @updateXRegression()
 
-        ($ "#regressionButton").click =>
+        $("#regressionButton").click =>
```
single quotes pease

kcarcia commented:
@JKinsman I need to pull and review tomorrow, but this code is so beautifully written. I found some console.logs, one small inconsistency, and it looks like you've might've missed a few of Jeremy's requests, but aside from that, great job.

jaypoulz commented:
@kcarcia He actually did exactly what I requested. I asked him to ignore a few of my requests because of merge issues. :smile:

kcarcia commented:
Oops, sorry @JKinsman! This looks fine to me. :)  @jaypoulz should I pull or should we have one other person review it?

jaypoulz commented:
I wanna talk to Jake before we pull this. Ill let you know when we're clear.

JKinsman commented:
@kcarcia haha, no worries!  It's totally okay :)

@jaypoulz, I'll be in Tuesday and Friday next week :)

On Fri, Apr 17, 2015 at 3:29 PM Jeremy Poulin <notifications@github.com>
wrote:

> I wanna talk to Jake before we pull this. Ill let you know when we're
> clear.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/isenseDev/rSENSE/pull/2096#issuecomment-94058250>.
>

jaypoulz commented:
@kcarcia Go ahead and pull this. What I have to talk about with Jake is related by not quintessential.
